zerver/lib/presence.py

```diff
@@ -35,72 +24,60 @@ def get_status_dicts_for_rows(
 
-    for user_key, presence_rows in itertools.groupby(all_rows, get_user_key):
+    for presence_row in all_rows:
+        user_key = get_user_key(presence_row)
         info = get_user_info(
-            list(presence_rows),
-            mobile_user_ids=mobile_user_ids,
+            presence_row["last_active_time"],
+            presence_row["last_connected_time"],
         )
         user_statuses[user_key] = info
 
     return user_statuses
 
 
 def get_modern_user_info(
-    presence_rows: Sequence[Mapping[str, Any]], mobile_user_ids: Set[int]
+    last_active_time: datetime.datetime, last_connected_time: datetime.datetime
 ) -> Dict[str, Any]:
-
-    active_timestamp = None
-    for row in reversed(presence_rows):
-        if row["status"] == UserPresence.ACTIVE:
-            active_timestamp = datetime_to_timestamp(row["timestamp"])
-            break
-
-    idle_timestamp = None
-    for row in reversed(presence_rows):
-        if row["status"] == UserPresence.IDLE:
-            idle_timestamp = datetime_to_timestamp(row["timestamp"])
-            break
-
-    # Be stingy about bandwidth, and don't even include
-    # keys for entities that have None values.  JS
-    # code should just do a falsy check here.
+    # TODO: Do further bandwidth optimizations to this structure.
     result = {}
-
-    if active_timestamp is not None:
-        result["active_timestamp"] = active_timestamp
-
-    if idle_timestamp is not None:
-        result["idle_timestamp"] = idle_timestamp
-
+    result["active_timestamp"] = datetime_to_timestamp(last_active_time)
+    result["idle_timestamp"] = datetime_to_timestamp(last_connected_time)
     return result
 
 
 def get_legacy_user_info(
-    presence_rows: Sequence[Mapping[str, Any]], mobile_user_ids: Set[int]
+    last_active_time: datetime.datetime, last_connected_time: datetime.datetime
 ) -> Dict[str, Any]:
-
-    # The format of data here is for legacy users of our API,
-    # including old versions of the mobile app.
-    info_rows = []
-    for row in presence_rows:
-        client_name = row["client__name"]
-        status = UserPresence.status_to_string(row["status"])
-        dt = row["timestamp"]
-        timestamp = datetime_to_timestamp(dt)
-        push_enabled = row["user_profile__enable_offline_push_notifications"]
-        has_push_devices = row["user_profile_id"] in mobile_user_ids
-        pushable = push_enabled and has_push_devices
-
-        info = dict(
-            client=client_name,
-            status=status,
-            timestamp=timestamp,
-            pushable=pushable,
-        )
-
-        info_rows.append(info)
-
-    most_recent_info = info_rows[-1]
+    # Reformats the modern UserPresence data structure so that legacy
+    # API clients can still access presence data.
+    #
+    # We expect this code to remain mostly unchanged until we can delete it.
+
+    if timezone_now() - last_active_time > datetime.timedelta(minutes=2):
```

Comment:
I think this `timedelta` may be wrong? Shouldn't this use the same value as
```
def format_legacy_presence_dict(presence: UserPresence) -> Dict[str, Any]:
    """
    This function assumes it's being called right after the presence object was updated,
    and is not meant to be used on old presence data.
    """
    if (
        presence.last_active_time + datetime.timedelta(minutes=1, seconds=10)
        presence.last_active_time
        + datetime.timedelta(seconds=settings.PRESENCE_LEGACY_EVENT_OFFSET_FOR_ACTIVITY_SECONDS)
        >= presence.last_connected_time
    ):
        status = UserPresence.LEGACY_STATUS_ACTIVE
        timestamp = datetime_to_timestamp(presence.last_active_time)
    else:
        status = UserPresence.LEGACY_STATUS_IDLE
        timestamp = datetime_to_timestamp(presence.last_connected_time)
```

which we've set to 70seconds?

Reply:
So I think the core problem statement here is the concern that you have multiple clients, and one is reporting IDLE but another is reporting ACTIVE, and we should not be overriding the "I'm active" information from the computer you're at with information from another client that thinks you're IDLE. (The new `last_connected_time`/`last_active_time` system is **much** cleaner on this point).

So I'm pretty sure the meaning of that 70s was basically "If you have a presence update showing you were IDLE within the last PRESENCE_UPDATE_MIN_FREQ_SECONDS + slop, but none showing you are ACTIVE, then we know there isn't another client that is reliably updating us with the information that you're active. 

I think if we assume that network failures don't happen, then indeed it doesn't matter whether we're doing this check just after the presence object was updated -- the threshold should always be.
`settings.PRESENCE_LEGACY_EVENT_OFFSET_FOR_ACTIVITY_SECONDS`. And I think it's pretty low harm for you to switch to being shown as IDLE (with a more current connection) if your computer fails to tell the server that you're online, so I don't think we need to add an extra round trip of slop here. 

Does that mean we can just call `format_legacy_presence_dict` here? I guess it's expecting a `UserPresence` object rather than two values, but that can probably be extracted.

Reply:
https://github.com/zulip/zulip/pull/16381/commits/464c33b1a671f9b5747c0c3618c4b69f2f332a41#r1157752227